### ISSUE TITLE
Chunk size is not always automatically found for fs.ReadStream

### ISSUE THREAD

**Issue by Hyzual** [tus/tus-js-client#454]
Hello, first of all thank you very much for this amazing library and specification, it's very easy to use it and works well! I have found a bug while using this library in a Node.js context.

**Describe the bug**
 The [node.js installation doc][0] says that if I pass an instance of `fs.ReadStream`, "no chunk will be held in memory". I took this to mean that I did not have to specify the `chunkSize` and `uploadSize` options. However, after trying to give it a stream from a [`fs.fileHandle`][1], `tus-js-client` will complain that it needs a finite `chunkSize` option. I am not sure if I am misunderstanding and it was always meant to work this way or if this is a case where `tus-js-client` did not correctly recognize what I'm giving it.

**To Reproduce**
I tried hard to setup a shareable reproduction with https://codesandbox.io but I did not succeed. They run Node.js v14 and `filehandle.createReadStream()` was only introduced in Node.js v16. I could not find a "code sandbox" website that allows me to run Node v16. If you know of any, please share it and I'll try to setup a reproduction there.

Here is the code I used on codesandbox. It creates a HTTP server on port 8080. When called with any request, it will read a `fixture.json` file (its contents don't matter) and upload it to `https://tusd.tusdemo.net/files/`. If you comment the `chunkSize` or `uploadSize` option, `tus-js-client` will throw an error explaining that it needs those options.

Steps:
1. Make sure you have Node.js v16.x
2. Create an `index.js` file and paste the code block below into it
3. Comment `chunkSize` option in `index.js`
4. Create a package.json and install `tus-js-client`
5. Create a `fixture.json` file next to `index.js` and write any JSON in it, for example `{"key": "value"}`
6. Run the following file: `node index.js`
7. Call the server: `curl localhost:8080`
8. It should respond "An error occurred" and you should see console error

```javascript
const http = require("http");
const fs = require("fs/promises");
const tus = require("tus-js-client");

const file_path = "./fixture.json"; // path to a file accessible to the node.js process. It will be uploaded to the TUS backend.
const upload_url = "https://tusd.tusdemo.net/files/"; // path to a remote TUS backend server
const CHUNK_SIZE = 67108864; // = 64 MiB. Number of bytes held in memory at a time during file upload;

function readFixtureAndUploadItToTUS() {
  return fs.open(file_path).then(async (handle) => {
    // I am forced to also run "stat" on the file to get its total size
    const file_stats = await handle.stat();

    return new Promise((resolve, reject) => {
      const uploader = new tus.Upload(handle.createReadStream({ start: 0 }), {
        endpoint: upload_url,
        // Note that tus-js-client's documentation specifies to avoid setting chunkSize and uploadSize unless forced to.
        // We are forced to set them, it does not look like the detection of ReadableStream worked in our case.
        chunkSize: CHUNK_SIZE,
        uploadSize: file_stats.size,
        onError: (error) => {
          handle.close();
          reject(error);
        },
        onSuccess: () => {
          handle.close();
          resolve();
        }
      });
      uploader.start();
    });
  });
}

http
  .createServer(function (req, res) {
    res.write("Starting the Upload !");
    readFixtureAndUploadItToTUS().then(
      () => {
        res.write("Successfully uploaded ! ");
        res.end();
      },
      (error) => {
        console.error(error);
        res.write("An error occurred");
        res.end();
      }
    );
  })
  .listen(8080);
```

**Expected behavior**
I expected **not** to have to set the `chunkSize` and `uploadSize` options so that `tus-js-client` would deduce them by itself.

**Setup details**
- Runtime environment: Node.js v16.17.0 (in Electron, specifically when developing a Visual Studio Code extension, if that makes any difference).
- Used tus-js-client version: 3.0.0
- Used tus server software: custom implementation in PHP, it does not seem related to the server software though.

[0]: https://github.com/tus/tus-js-client/blob/master/docs/installation.md#nodejs-support
[1]: https://nodejs.org/docs/latest-v16.x/api/fs.html#class-filehandle

**Comment by Acconut** [comment]
Thanks for the detailed report. tus-js-client currently relies on the `path` property of the `ReadStream` class: https://github.com/tus/tus-js-client/blob/14c3634e0ccd45973e90a3e460ca5e749f630107/lib/node/fileReader.js#L14

It is used to get the file size via `stat`: https://github.com/tus/tus-js-client/blob/14c3634e0ccd45973e90a3e460ca5e749f630107/lib/node/sources/FileSource.js#L6

Since you get it from an file descriptor, this property is not present and tus-js-client falls back to handling it as any arbitrary readable stream, which requires buffer memory and the `chunkSize` option.

I think in theory we could use `fstat` to get the size via the file descriptor, if it is present but I haven't tested that yet. Would you be interested in submitting a PR?

**Comment by Hyzual** [comment]
Thanks for the very quick answer ! Indeed maybe `fstat` could be used to auto-detect the file size. I'm sorry, I don't think I will have much time to commit to getting into the codebase and submitting a PR. I might try it on my free time though.

**Comment by Acconut** [comment]
> I'm sorry, I don't think I will have much time to commit to getting into the codebase and submitting a PR. I might try it on my free time though.

I do understand that! I am also not sure when I have time to address this.

**Comment by major-winter** [comment]
Hello everyone, this issue is still open, and I’d like to take a shot at it. I was able to reproduce the bug, and I’d appreciate some clarification from you, @Acconut.

The issue seems to occur because a `ReadStream` object is expected to have an associated `path property`,  which isn’t the case here. As a result, `chunkSize` should be explicitly provided in the options. From what I can tell, the original code uses a `ReadStream` object rather than a file descriptor. This is confusing because your comment mentioned using `fstat` to determine the file size from a descriptor.

I propose that we provide a default `chunkSize` when the `ReadStream` instance does not have a path property. What do you think?

**Comment by Acconut** [comment]
An `fs.ReadStream` can either be based on file path, when it's created using `fs.createReadStream('path/to/my/file.txt')`. The `fs.ReadStream` instance then has a `path` property that tus-js-client can use to determine the file size.

However, a `fs.ReadStream` can also be created from a file descriptor, for example using `filehandle.createReadStream()`. In that case, the `fs.ReadStream` instance has no `path` property that we can run `stat` against. Instead, it has a `fd` property that we can use `fstat` on. At least in theory, I haven't tried this.

If you are running into the same problem as @Hyzual, I would suggest using `fs.createReadStream` directly instead of `filehandle.createReadStream`. Then, tus-js-client will accept the object without you having to specify a chunkSize or uploadSize.

> I propose that we provide a default `chunkSize` when the `ReadStream` instance does not have a path property. What do you think?

I don't think that's a good approach. Buffering a readable stream locally will consume memory and users of tus-js-client should explicitly opt into this behavior. This might be fine for one upload, but when handling many uploads in parallel, it can lead to OOM issues. With a default value for chunkSize, tus-js-client will consume resources that are hard to understand for users as they never agreed to this buffering.

Does that make sense?

**Comment by Acconut** [comment]
This discussion here got me thinking that it's odd that tus-js-client requires users to create an fs.ReadStream when they want to upload a file on disk without reading it into memory. While it might make sense in the first place, tus-js-client actually doesn't consume data via the provided fs.ReadStream but instead opens its own fs.ReadStream depending on the offset that it currently needs to read data from. At the end of the day, the original fs.ReadStream instance is basically useless.

That's why I think it's better to allow users to pass paths directly to tus-js-client, which it will then upload without needing the user to open fs.ReadStreams. I've started working on this in https://github.com/tus/tus-js-client/pull/767 and it seems to be working well. By allowing basically any object with a `path` property to be recognized by tus-js-client, we even keep backwards compatibility as fs.ReadStreams satisfy this interface.

The reason why I bring this up here is that with this change, users will be discouraged to pass fs.ReadStreams to tus-js-client as it's more efficient to just give it the path. Then we also don't have to add support for fs.ReadStreams that were created using file handles because there is an easier alternative for users. In addition, I have experimented with the file handle API and integrating it into tus-js-client might not be trivial because there is one file descriptor and tus-js-client and the user have to be very careful so that this one file descriptor isn't closed or otherwise the upload will fail.

Overall, I'm considering leaving this ticket without further action and instead going the route of https://github.com/tus/tus-js-client/pull/767, where the tus-js-client API is extended to not require the odd way through fs.ReadStream at all anymore.

What do you think? Does that make sense?

**Comment by major-winter** [comment]
I agree. The provided `fs.ReadStream` is only used to determine the offset and does not make sense for the client to send a ReadStream instance at all. I believe this ticket can be closed in favor of ticket #767. Further discussion can be done in that ticket.

**Comment by Acconut** [comment]
I agree, thank you for your feedback. Let's go with #767
